src/QUICConnection.ts

```diff
@@ -515,505 +549,762 @@ class QUICConnection extends EventTarget {
+        const e = new errors.ErrorQUICConnectionInternal(
+          'Failed processing stream, stream was writable even though `QUICStream` does not exist',
+        );
+        // this.conn.close(
+        //   false,
+        //   ConnectionErrorCode.InternalError,
+        //   Buffer.from('')
+        // );
+        // this.dispatchEvent(new events.EventQUICConnectionError({ detail: e }));
+        // this.dispatchEvent(new events.EventQUICConnectionClose(
+        //   {
+        //     detail: {
+        //       type: 'local',
+        //       ...this.conn.localError()!
+        //     }
+        //   }
+        // ));
+        // This would be a software error
+        throw e;
       } else {
         quicStream.write();
       }
     }
   }
 
-  protected setConnTimeOutTimer(): void {
-    const logger = this.logger.getChild('timer');
-    const connTimeOutHandler = async () => {
+  /**
+   * This only gets called on the first `send`.
+   * Aftewards, nothing can cancel this, except here.
+   * It's the responsiblity of this timer to resolve the `closedP` normally.
+   * This is because when a quiche connection is closing, it enters a draining phase
+   * and stays draining for some time, before finally becoming closed.
+   */
+  protected setConnTimeoutTimer(): void {
+    const connTimeoutHandler = async (signal: AbortSignal) => {
+      // If aborted, just immediately resolve
+      if (signal.aborted) return;
       // This can only be called when the timeout has occurred.
       // This transitions the connection state.
       // `conn.timeout()` is time aware, so calling `conn.onTimeout` will only
       //  trigger state transitions after the time has passed.
-      logger.debug('CALLING ON TIMEOUT');
       this.conn.onTimeout();
-
-      // Connection may have closed after timing out
+      // If it is closed, we can resolve, and we are done for this connection
+      // So we can just return, nothing more to do
       if (this.conn.isClosed()) {
-        // If it was still starting waiting for the secure event, we need to
-        //  reject the `secureEstablishedP` promise.
-        if (this[status] === 'starting') {
-          this.rejectSecureEstablishedP(
-            new errors.ErrorQUICConnectionInternal('Connection has closed!'),
-          );
-        }
-
-        logger.debug('resolving closedP');
-        // We resolve closing here, stop checks if the connection has timed out
-        //  and handles it.
         this.resolveClosedP();
-        // If we are still running and not stopping then we need to stop
-        if (this[running] && this[status] !== 'stopping') {
-          // Background stopping, we don't want to block the timer resolving
-          void this.stop({ force: true });
-        }
-        logger.debug('CLEANING UP TIMER');
         return;
       }
-
-      // There may be data to send after timing out
-      void this.send();
-
+      // Otherwise, we should be calling send after the timeout
```

Comment:
We shouldn't return after closing here. If it closed due to the connections timing out from the `maxIdleTimeout` then we need to dispatch the error and close event.

Reply:
Oh wow, the diff is really bad here. In any case, we should be checking if the connection timed out here rather than later.

Reply:
What was this?

Reply:
Done in 5616354d15ab9e020a12b86598c842ee21d8b170